File: tests/test_residual.py
import os
import numpy as np
from nose.tools import assert_almost_equals
import simpl
import simpl.peak_detection as peak_detection
import simpl.partial_tracking as partial_tracking
import simpl.synthesis as synthesis
import simpl.residual as residual

float_precision = 2
frame_size = 512
hop_size = 512
max_peaks = 10
max_partials = 10
num_frames = 30
num_samples = num_frames * hop_size
audio_path = os.path.join(
    os.path.dirname(__file__), 'audio/flute.wav'
)
libsms_residual_synthesis_path = os.path.join(
    os.path.dirname(__file__), 'libsms_residual_synthesis.wav'
)

PeakDetection = peak_detection.PeakDetection
SMSPeakDetection = peak_detection.SMSPeakDetection
PartialTracking = partial_tracking.PartialTracking
SMSPartialTracking = partial_tracking.SMSPartialTracking
Synthesis = synthesis.Synthesis
SMSSynthesis = synthesis.SMSSynthesis
Residual = residual.Residual
SMSResidual = residual.SMSResidual


class TestResidual(object):
    @classmethod
    def setup_class(cls):
        cls.audio = simpl.read_wav(audio_path)[0]

    def test_basic(self):
        pd = PeakDetection()
        frames = pd.find_peaks(self.audio)

        pt = PartialTracking()
        frames = pt.find_partials(frames)

        synth = Synthesis()
        synth_audio = synth.synth(frames)

        res = Residual()
        residual_audio = res.find_residual(synth_audio, self.audio)
        assert len(residual_audio)


class TestSMSResidual(object):
    @classmethod
    def setup_class(cls):
        cls.audio = simpl.read_wav(audio_path)[0]
        cls.audio = cls.audio[0:num_samples]

    def test_basic(self):
        pd = SMSPeakDetection()
        pd.hop_size = hop_size
        frames = pd.find_peaks(self.audio)

        pt = SMSPartialTracking()
        pt.max_partials = max_partials
        frames = pt.find_partials(frames)

        synth = SMSSynthesis()
        synth.hop_size = hop_size
        synth_audio = synth.synth(frames)

        res = SMSResidual()
        residual_audio = res.find_residual(synth_audio, self.audio)
        assert len(residual_audio)

    def test_residual_synthesis(self):
        res = SMSResidual()
        res.hop_size = hop_size
        simpl_residual = res.synth(self.audio)

        sms_residual, sampling_rate = simpl.read_wav(
            libsms_residual_synthesis_path
        )

        assert len(simpl_residual) == len(sms_residual)

        import matplotlib.pyplot as plt
        plt.plot(simpl_residual)
        plt.plot(sms_residual)
        plt.show()

        for i in range(len(simpl_residual)):
            assert_almost_equals(simpl_residual[i], sms_residual[i],
                                 float_precision)

File: tests/test_residual.cpp
#include "test_residual.h"

using namespace simpl;

// ---------------------------------------------------------------------------
//	test_basic
// ---------------------------------------------------------------------------
static void test_basic(Residual* residual, SndfileHandle *sf) {
    int num_samples = 4096;
    int hop_size = 256;
    int frame_size = 512;

    std::vector<sample> audio(sf->frames(), 0.0);
    sf->read(&audio[0], (int)sf->frames());

    residual->reset();
    residual->frame_size(frame_size);
    residual->hop_size(hop_size);

    Frames frames = residual->synth(num_samples,
                                    &(audio[(int)sf->frames() / 2]));

    for(int i = 0; i < frames.size(); i++) {
        CPPUNIT_ASSERT(frames[i]->num_peaks() > 0);
        CPPUNIT_ASSERT(frames[i]->num_partials() > 0);

        double energy = 0.f;
        for(int j = 0; j < residual->hop_size(); j++) {
            energy += frames[i]->synth_residual()[j] *
                      frames[i]->synth_residual()[j];
        }
        CPPUNIT_ASSERT(energy > 0.f);
    }
}


// ---------------------------------------------------------------------------
//	TestSMSResidual
// ---------------------------------------------------------------------------
void TestSMSResidual::setUp() {
    _sf = SndfileHandle(TEST_AUDIO_FILE);

    if(_sf.error() > 0) {
        throw Exception(std::string("Could not open audio file: ") +
                        std::string(TEST_AUDIO_FILE));
    }
}

void TestSMSResidual::test_basic() {
    ::test_basic(&_res, &_sf);
}
